plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,484 @@
+import           Plutus.SCB.Utils                                  (render, tshow)
+
+import qualified Plutus.SCB.Core.Projections                       as Projections
+
+sendContractStateMessages ::
+    forall t effs.
+        ( Member (EventLogEffect (ChainEvent t)) effs
+        , Member (Error SCBError) effs
+        , Member Log effs
+        )
+    => t
+    -> ContractInstanceId
+    -> ContractIteration
+    -> PartiallyDecodedResponse
+    -> Eff effs ()
+sendContractStateMessages t instanceID iteration newState = do
+    logInfo . render $ "Sending messages for contract" <+> pretty instanceID <+> "at iteration" <+> pretty iteration
+    let is = ContractInstanceState
+            { csContract = instanceID
+            , csCurrentIteration = iteration
+            , csCurrentState = newState
+            , csContractDefinition = t
+            }
+    void
+        $ runCommand (sendContractEvent @t) ContractEventSource
+        $ ContractInstanceStateUpdateEvent
+        $ is
+
+    void $ runCommand (saveContractState @t) UserEventSource is
+
+    decodedRequests <- Contract.decodeContractRequests newState
+    logInfo . render $ "There are" <+> pretty (length decodedRequests) <+> "requests for this contract"
+    traverse_ (sendContractOutboxMessage @t instanceID iteration) decodedRequests
+
+sendContractOutboxMessage ::
+    forall t effs.
+        ( Member (EventLogEffect (ChainEvent t)) effs
+        )
+    => ContractInstanceId
+    -> ContractIteration
+    -> ContractRequest
+    -> Eff effs [ChainEvent t]
+sendContractOutboxMessage instanceID iteration msg =
+    runCommand (sendContractEvent @t) ContractEventSource
+        $ ContractOutboxEvent
+        $ ContractOutboxMessage
+            { coInstance = instanceID
+            , coIteration = iteration
+            , coPayload = msg
+            }
+
+sendContractInboxMessage ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    )
+    => ContractResponse
+    -> ContractInstanceId
+    -> ContractIteration
+    -> Eff effs [ChainEvent t]
+sendContractInboxMessage rsp i it = do
+    runCommand (sendContractEvent @t) ContractEventSource
+        $ ContractInboxEvent
+        $ ContractInboxMessage
+            { ciInstance = i
+            , ciIteration = it
+            , ciPayload = rsp
+            }
+
+-- | Get the current state of the contract instance
+lookupContractState ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member (Error SCBError) effs
+    )
+    => ContractInstanceId
+    -> Eff effs (ContractInstanceState t)
+lookupContractState instanceID = do
+    mp <- runGlobalQuery (Query.contractState @t)
+    case Map.lookup instanceID mp of
+        Nothing       -> throwError $ OtherError $ "Contract instance not found " <> tshow instanceID
+        Just (Last s) -> pure s
+
+-- | For a given contract instance, take the first message
+--   from the inbox and update the contract with it.
```

Comment:
I was confused reading the code because I'd assumed it would process the whole inbox. Maybe we can name this more clearly...

Reply:
I changed it so that it's clear that it only processes one message.